src/Mod/Part/Gui/PartParams.cpp

```diff
@@ -0,0 +1,102 @@
+/****************************************************************************
+ *   Copyright (c) 2020 Zheng, Lei (realthunder) <realthunder.dev@gmail.com>*
+ *                                                                          *
+ *   This file is part of the FreeCAD CAx development system.               *
+ *                                                                          *
+ *   This library is free software; you can redistribute it and/or          *
+ *   modify it under the terms of the GNU Library General Public            *
+ *   License as published by the Free Software Foundation; either           *
+ *   version 2 of the License, or (at your option) any later version.       *
+ *                                                                          *
+ *   This library  is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of         *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the          *
+ *   GNU Library General Public License for more details.                   *
+ *                                                                          *
+ *   You should have received a copy of the GNU Library General Public      *
+ *   License along with this library; see the file COPYING.LIB. If not,     *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,          *
+ *   Suite 330, Boston, MA  02111-1307, USA                                 *
+ *                                                                          *
+ ****************************************************************************/
+
+#include "PreCompiled.h"
+#include <App/Application.h>
+#include <App/Document.h>
+#include <Gui/Application.h>
+#include <Gui/Document.h>
+#include "ViewProvider.h"
+#include "PartParams.h"
+
+using namespace PartGui;
+
+PartParams::PartParams() {
+    handle = App::GetApplication().GetParameterGroupByPath(
+            "User parameter:BaseApp/Preferences/Mod/Part");
+    handle->Attach(this);
+#undef FC_PART_PARAM
```

Comment:
why undef? it seems to me it could suppress potentially valuable "macro redefinition" warnings...

Reply:
This is an essential step of a very useful Macro usage pattern. Take `PartParam` for an example. I defined the main parameter list macro `FC_PART_PARAMS`, consists a list of entry of 

```
FC_PART_PARAM(_name, _ctype, _type, _def)
```

Each entry corresponds to a parameter providing necessary information to generate various type of code. For example, it can generate class member declaration

```
#define FC_PART_PARAM(_name, _ctype, _type, _def) \
    _ctype _##_name;
````

Generate code for setter and getter function

```
#define FC_PART_PARAM(_name,_ctype,_type,_def) \
    static const _ctype & _name() { return instance()->_##_name; }\
    static void set_##_name(_ctype _v) { instance()->handle->Set##_type(#_name,_v); instance()->_##_name=_v; }
```

Generate code to initialize parameter value from parameter group,

```
#define FC_PART_PARAM(_name, _ctype, _type, _def) \
    this->_##_name = handle->Get##_type(#_name,_def);
````

The point is that, by `undef` and re-`define` the same `FC_PART_PARAM` macro, I can easily generate the same repetitive code for all parameters. Adding a parameter is to simply add a new entry in `FC_PART_PARAMS`.

IMO, this usage pattern is particular useful for managing parameters, settings and stuff. For example, I used it to generate code for preference setting page [here](https://github.com/realthunder/FreeCAD/blob/04c80c99eee0f7d2f761bdc7d8b32c077adefdec/src/Gui/DlgSettingsUI.cpp#L47). These settings come from [Gui::ViewParams](https://github.com/realthunder/FreeCAD/blob/04c80c99eee0f7d2f761bdc7d8b32c077adefdec/src/Gui/ViewParams.h#L49), which is similar to `PartParams` here, but with additional documentation strings embedded.